cpp/src/arrow/python/python_to_arrow.cc

```diff
@@ -0,0 +1,646 @@
+    ADD_ELEMENT(doubles_, double_tag);
+
+    ADD_ELEMENT(tensor_indices_, tensor_tag);
+
+    ADD_SUBSEQUENCE(list_data, list_offsets_, list_builder, list_tag, "list");
+    ADD_SUBSEQUENCE(tuple_data, tuple_offsets_, tuple_builder, tuple_tag, "tuple");
+    ADD_SUBSEQUENCE(dict_data, dict_offsets_, dict_builder, dict_tag, "dict");
+
+    auto type = ::arrow::union_(types, type_ids, UnionMode::DENSE);
+    out->reset(new UnionArray(type, types_.length(), children, types_.data(),
+                              offsets_.data(), nones_.null_bitmap(),
+                              nones_.null_count()));
+    return Status::OK();
+  }
+
+ private:
+  MemoryPool* pool_;
+
+  Int8Builder types_;
+  Int32Builder offsets_;
+
+  NullBuilder nones_;
+  BooleanBuilder bools_;
+  Int64Builder ints_;
+  BinaryBuilder bytes_;
+  StringBuilder strings_;
+  FloatBuilder floats_;
+  DoubleBuilder doubles_;
+
+  // We use an Int32Builder here to distinguish the tensor indices from
+  // the ints_ above (see the case Type::INT32 in get_value in python.cc).
+  // TODO(pcm): Replace this by using the union tags to distinguish between
+  // these two cases.
+  Int32Builder tensor_indices_;
+
+  std::vector<int32_t> list_offsets_;
+  std::vector<int32_t> tuple_offsets_;
+  std::vector<int32_t> dict_offsets_;
+
+  // Tags for members of the sequence. If they are set to -1 it means
+  // they are not used and will not part be of the metadata when we call
+  // SequenceBuilder::Finish. If a member with one of the tags is added,
+  // the associated variable gets a unique index starting from 0. This
+  // happens in the UPDATE macro in sequence.cc.
+  int8_t bool_tag = -1;
+  int8_t int_tag = -1;
+  int8_t string_tag = -1;
+  int8_t bytes_tag = -1;
+  int8_t float_tag = -1;
+  int8_t double_tag = -1;
+
+  int8_t tensor_tag = -1;
+  int8_t list_tag = -1;
+  int8_t tuple_tag = -1;
+  int8_t dict_tag = -1;
+
+  int8_t num_tags = 0;
+};
+
+/// Constructing dictionaries of key/value pairs. Sequences of
+/// keys and values are built separately using a pair of
+/// SequenceBuilders. The resulting Arrow representation
+/// can be obtained via the Finish method.
+class DictBuilder {
+ public:
+  explicit DictBuilder(MemoryPool* pool = nullptr) : keys_(pool), vals_(pool) {}
+
+  /// Builder for the keys of the dictionary
+  SequenceBuilder& keys() { return keys_; }
+  /// Builder for the values of the dictionary
+  SequenceBuilder& vals() { return vals_; }
+
+  /// Construct an Arrow StructArray representing the dictionary.
+  /// Contains a field "keys" for the keys and "vals" for the values.
+
+  /// \param list_data
+  ///    List containing the data from nested lists in the value
+  ///   list of the dictionary
+  ///
+  /// \param dict_data
+  ///   List containing the data from nested dictionaries in the
+  ///   value list of the dictionary
+  Status Finish(std::shared_ptr<Array> key_tuple_data,
+                std::shared_ptr<Array> key_dict_data,
```

Comment:
Make the shared_ptr arguments a constant reference, this should avoid needless copies